ExpressionUtils/Evaluating/CachedExpressionCompiler.cs

```diff
@@ -82,7 +83,29 @@ private static object getDefaultValue(Type t) {
 		/// Use for testing only.
 		/// </remarks>
 		internal bool IsCached(LambdaExpression lambda) {
-			return delegates.ContainsKey(lambda);
+			return delegates.ContainsKey((lambda.Body, lambda.Parameters));
+		}
+
+		/// <summary>
+		/// Previously as a key for <see cref="delegates"/> we were using <see cref="LambdaExpression"/> objects and <see cref="ExpressionComparing.StructuralComparer"/> as a comparer.
+		/// But that required us to make calls to <see cref="Expression.Lambda(Expression, ParameterExpression[])"/> which contains global lock inside, and that started to become a problem.
+		/// So instead we now use <see cref="(Expression , IReadOnlyCollection{ParameterExpression})"/> tuple as a key and pass body and parameters separately, to reduce number of calls to <see cref="Expression.Lambda(Expression, ParameterExpression[])"/>.
+		/// </summary>

```

Comment:
Maybe this summary fits more at the definition of the `delegates``dictionary at the top?

Reply:
I think its better to have this description for `LambdaExpressionComparer` since that cache key change is the only reason for its existence. While `delegates` dictionary plays the same role whether you use tuple of `(body, params)` or lambda as its key.

But i can change this if you insist.

Reply:
Can you mention `LambdaExpressionComparer` in this comment to make it more explicit that this is the explanation why we need that comparer? (rather than the explanation why we need the cache key tuple, which I think would fit more at the declaration of the delegates dictionary)